src/enzyme_ad/jax/Dialect/EnzymeXLAOps.td

```diff
@@ -1058,4 +1058,141 @@ def AffineStoreVar : EnzymeXLA_Op<"store_var", [Pure]> {
   let summary = "Fake store an SSA value for conversion to ISL";
 }
 
+// MPI Ops
+
+def MPICommRankOp : EnzymeXLA_Op<"comm_rank", []> {
+  let summary = "Equivalent to " "`MPI_Comm_rank(MPI_COMM_WORLD, &rank)`";
+
+  let arguments = (
+    ins AnyTensor : $inrank
+  );
+
+  let results = (
+    outs TensorOf<[I32]> : $outrank
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` functional-type(operands, results)";
+}
+
+def MPICommSizeOp : EnzymeXLA_Op<"comm_size", []> {
+  let summary = "Equivalent to MPI_Comm_size(MPI_COMM_WORLD, &size)";
+
+  let arguments = (
+    ins AnyTensor : $insize
+  );
+
+  let results = (
+    outs TensorOf<[I32]> : $outsize
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` functional-type(operands, results)";
+}
+
+// We call it "mpi_barrier" to not collide with BarrierOp above
+def MPIBarrierOp : EnzymeXLA_Op<"mpi_barrier", []> {
+  let summary = "Equivalent to MPI_Barrier(MPI_COMM_WORLD)";
+  let assemblyFormat = "attr-dict";
+}
+
+def MPISendOp : EnzymeXLA_Op<"send", []> {
+  let summary = "Equivalent to "
+                "`MPI_Send(&buf, count, datatype, dest, tag, comm)`";
+
+  let arguments = (
+    ins AnyTensor : $buf,
+    TensorOf<[I32]> : $count,
+    TensorOf<[I32]> : $dest,
+    TensorOf<[I32]> : $tag,
+    StrAttr:$datatype
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` type(operands)";
+}
+
+def MPIRecvOp : EnzymeXLA_Op<"recv", []> {
+  let summary = "Equivalent to "
+                "`MPI_Recv(&buf, count, datatype, source, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE)`";
+
+  let arguments = (
+    ins AnyTensor : $inbuf,
+    TensorOf<[I32]> : $count,
+    TensorOf<[I32]> : $source,
+    TensorOf<[I32]> : $tag,
+    StrAttr:$datatype
+  );
+
+  let results = (
+    outs AnyTensor : $outbuf
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` functional-type(operands, results)";
+}
+
+def MPIIsendOp : EnzymeXLA_Op<"isend", []> {
+  let summary = "Equivalent to "
+                "`MPI_Isend(&buf, count, datatype, dest, tag, MPI_COMM_WORLD, &request)`";
+
+  let arguments = (
+    ins AnyTensor : $buf,
+    TensorOf<[I32]> : $count,
+    TensorOf<[I32]> : $dest,
+    TensorOf<[I32]> : $tag,
+    TensorOf<[I64]> : $inrequest,
+    StrAttr:$datatype
+  );
+
+  let results = (
+    outs TensorOf<[I64]> : $outrequest
+  );
+
+  let assemblyFormat = "`(` operands `)` attr-dict `:` functional-type(operands, results)";
+}
+
+def MPIIrecvOp : EnzymeXLA_Op<"irecv", []> {
+  let summary = "Equivalent to "
+                "`MPI_Irecv(&buf, count, datatype, source, tag, MPI_COMM_WORLD, &request)`";
+
+  let arguments = (
+    ins AnyTensor : $inbuf,
+    TensorOf<[I32]> : $count,
+    TensorOf<[I32]> : $source,
+    TensorOf<[I32]> : $tag,
+    TensorOf<[I64]> : $inrequest,
+    StrAttr:$datatype
+  );
+
+  let results = (
+    outs AnyTensor : $outbuf,
+    TensorOf<[I64]> : $outrequest
```

Comment:
I'm unsure about this `inbuf`/`outbuf` `inrequest`/`outrequest` design here and in other Ops. It allows us to reproduce exactly the same IR that we had before (ie, via manual lowering from Reactant). But the more I've thought about this the more it seems like a weird design for certain Ops/arguments. Eg, here it maybe makes sense for the buffers, but not the requests? And for `Comm_rank`, it seems like it would make more sense for it to be a pure function, where we don't take in a rank, only output one. However, doing that would require some changes to the way we call this on the Reactant side.

Reply:
this should only return a request, not take as an input [since its write only]